{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hidden Markov Models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generalized Viterbi Decoding Algorithm\n",
    "- states --> tuple of strings represeting sigma, the set of all possible states\n",
    "- obs --> the observed sequence of emission characters, represents Q\n",
    "- s_pro --> starting probability dictionary that maps each state to a probability\n",
    "- t_pro --> transition probability that maps each state to another dict that maps each state to a probability, represents A\n",
    "- e_pro --> emission probability that maps each state to another dict that maps each emission character to a probability, represents E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Viterbi(obs, states, s_pro, t_pro, e_pro):\n",
    "    path = { s:[] for s in states}\n",
    "    curr_pro = {}\n",
    "    for s in states:\n",
    "        curr_pro[s] = s_pro[s]*e_pro[s][obs[0]]\n",
    "    for i in range(1, len(obs)):\n",
    "        last_pro = curr_pro\n",
    "        curr_pro = {}\n",
    "        for curr_state in states:\n",
    "            max_pro, last_sta = max(((last_pro[last_state]*t_pro[last_state][curr_state]*e_pro[curr_state][obs[i]], last_state) \n",
    "                       for last_state in states))\n",
    "            curr_pro[curr_state] = max_pro\n",
    "            path[curr_state].append(last_sta)\n",
    "\n",
    "    # backtrack to find max path\n",
    "    max_pro = -1\n",
    "    max_path = None\n",
    "    for s in states:\n",
    "        path[s].append(s)\n",
    "        if curr_pro[s] > max_pro:\n",
    "            max_path = path[s]\n",
    "            max_pro = curr_pro[s]\n",
    "    return (max_path, max_pro)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mixed Genome Problem\n",
    "- given a set of alleles from the alphabet {A,C,G,T,H,N} (where H and N indicate probe errors) from 8 founder genomes and a 9th hybrid genome, determine the founder genome that most likely contributed to each allele in the hybrid genome\n",
    "    - i.e. determine the most likely sequence of mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read the Genome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Chromosome', 'Position', 'A/J', 'C57BL/6J', '129S1/SvImJ', 'NOD/ShiLtJ', 'NZO/HlLtJ', 'CAST/EiJ', 'PWK/PhJ', 'WSB/EiJ', 'owenmc']\n",
      "['13', 3561535, 'A', 'C', 'A', 'C', 'C', 'C', 'C', 'C', 'A']\n",
      "['13', 4029227, 'A', 'A', 'A', 'A', 'G', 'G', 'A', 'A', 'G']\n",
      "['13', 4752533, 'C', 'T', 'C', 'T', 'C', 'T', 'C', 'T', 'C']\n",
      "['13', 5392085, 'T', 'G', 'T', 'T', 'G', 'T', 'G', 'T', 'T']\n",
      "['13', 5945123, 'G', 'G', 'G', 'A', 'G', 'A', 'G', 'G', 'A']\n",
      "['13', 6304055, 'G', 'A', 'G', 'G', 'G', 'G', 'G', 'G', 'G']\n",
      "['13', 7088753, 'T', 'T', 'T', 'C', 'C', 'T', 'C', 'T', 'C']\n",
      "['13', 7553047, 'A', 'A', 'A', 'C', 'C', 'A', 'C', 'C', 'C']\n",
      "['13', 7999300, 'G', 'G', 'G', 'T', 'T', 'G', 'T', 'G', 'T']\n",
      "['13', 8618531, 'T', 'T', 'T', 'C', 'C', 'C', 'C', 'C', 'C']\n"
     ]
    }
   ],
   "source": [
    "fp = open(\"GenomeData.csv\", 'r')\n",
    "data = fp.read().split('\\n')        # break file into lines\n",
    "fp.close()\n",
    "header = data.pop(0).split(',')     # First line is header\n",
    "column = header.index(\"owenmc\")\n",
    "# while (len(data[-1].strip()) < 1):  # remove extra lines\n",
    "#     data.pop()\n",
    "for i, line in enumerate(data):     # make a list from each row\n",
    "    field = line.split(',')\n",
    "    field[1] = int(field[1])        # convert position to integer\n",
    "    for base in field[2:10]:\n",
    "        if base != field[3]:\n",
    "            break\n",
    "    data[i] = field[0:10] + [base]\n",
    "    \n",
    "print(header[0:10] + [header[column]])\n",
    "\n",
    "for i in range(0, 10):\n",
    "    print((data[i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Viterbi Decoding Algorithm Dynamic Program"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of paths: 261\n"
     ]
    }
   ],
   "source": [
    "from math import exp, log10\n",
    "\n",
    "Nstates = 8\n",
    "prevpos = 1\n",
    "state = [[(float(len(data)),i) for i in range(Nstates)]]   # (log(p), PathToHere)\n",
    "for i in range(len(data)):\n",
    "    # Count expected genotypes\n",
    "    count = dict([(call, data[i][2:2+Nstates].count(call)) for call in \"ACGTHN\"])\n",
    "    # Get the target genotype at this probe\n",
    "    observed = data[i][-1]\n",
    "    # Compute emission probability, assuming 5% error rate\n",
    "    if (count[observed] == 0):\n",
    "        emission = [1.0/Nstates for j in xrange(2,2+Nstates)]   # unexpected \n",
    "    else:\n",
    "        emission = [0.99/count[data[i][j]] if data[i][j] == observed else 0.01/(Nstates - count[data[i][j]])\n",
    "                    for j in range(2,2+Nstates)]\n",
    "    # compute transition probability\n",
    "    position = data[i][1]\n",
    "    delta = position - prevpos\n",
    "    prevpos = position\n",
    "    stay = ((Nstates - 1.0)*exp(-delta/23604450.50) + 1.0)/Nstates\n",
    "    switch = (1.0 - stay)/(Nstates - 1.0)\n",
    "    # update state probailities for all paths leading to the ith state\n",
    "    path = []\n",
    "    for j in range(Nstates):\n",
    "        choices = [(log10(emission[j])+(log10(stay) if (k==j) else log10(switch))+state[-1][k][0],k)\n",
    "                   for k in range(Nstates)]\n",
    "        path.append(max(choices))   # choices is a list of tuples of (score[i], from_whence_I_arrived[i])\n",
    "    state.append(path)\n",
    "print(\"Length of paths:\", len(state))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Backtrack to find solution: position, allele, most likely founder genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 (-13.933684392229354, 2) 129S1/SvImJ\n",
      "3561535 A NZO/HlLtJ\n",
      "4029227 G NZO/HlLtJ\n",
      "4752533 C NZO/HlLtJ\n",
      "5392085 T NOD/ShiLtJ\n",
      "5945123 A NOD/ShiLtJ\n",
      "6304055 G NOD/ShiLtJ\n",
      "7088753 C NOD/ShiLtJ\n",
      "7553047 C NOD/ShiLtJ\n",
      "7999300 T NOD/ShiLtJ\n",
      "8618531 C NOD/ShiLtJ\n"
     ]
    }
   ],
   "source": [
    "# backtrack\n",
    "path = state[-1]\n",
    "maxi = 0\n",
    "maxp = path[0][0]\n",
    "for i in range(1,Nstates):\n",
    "    if (path[i][0] > maxp):\n",
    "        maxp = path[i][0]\n",
    "        maxi = i\n",
    "print(maxi, path[maxi], header[2+maxi])\n",
    "\n",
    "for j in range(len(state)-2,-1,-1):\n",
    "    data[j].append(header[2+maxi])\n",
    "    maxi = state[j+1][maxi][1]\n",
    "\n",
    "# print the first 10 rows\n",
    "for row in data[:10]:\n",
    "    print(row[1], row[-2], row[-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fair Bet Casino Problem\n",
    "- Î£ = {H, T}\n",
    "- Q = {F, B}\n",
    "- A = [[0.9, 0.1],[0.1, 0.9]]\n",
    "- E = [[0.5, 0.5],[0.25,0.75]]\n",
    "\n",
    "- the unfair casino has fair and biased coins. The dealer swaps coins after each flip with probability of 0.1. The fair coin has 50-50 odds. The biased coin has 0.75, 0.25 odds for heads and tails, respectively.\n",
    "\n",
    "- given a sequence of coin flips, use a hmm to determine the most likely sequece of coins used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = \"HTHHHHHHHHHTHTTTHTHTHHHHTTHHTHTTTHHTHHHHHHHTHHHTHHTTTHHHHTTHHHHHHHTTHTHHHHHHHTTH\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ('F', 'B')\n",
    " \n",
    "observations = ('H', 'T')\n",
    " \n",
    "start_probability = {'F': 0.5, 'B': 0.5}\n",
    " \n",
    "transition_probability = {\n",
    "   'F' : {'F': 0.9, 'B': 0.1},\n",
    "   'B' :   {'B': 0.9, 'F': 0.1},\n",
    "   }\n",
    " \n",
    "emission_probability = {\n",
    "   'F' : {'H': 0.5, 'T': 0.5},\n",
    "   'B'   : {'H': 0.75, 'T': 0.25}\n",
    "   }\n",
    "\n",
    "\n",
    "def Viterbi(obs, states, s_pro, t_pro, e_pro):\n",
    "    path = { s:[] for s in states} # init path: path[s] represents the path ends with s\n",
    "    curr_pro = {}\n",
    "    for s in states:\n",
    "        curr_pro[s] = s_pro[s]*e_pro[s][obs[0]]\n",
    "    for i in range(1, len(obs)):\n",
    "        last_pro = curr_pro\n",
    "        curr_pro = {}\n",
    "        for curr_state in states:\n",
    "            max_pro, last_sta = max(((last_pro[last_state]*t_pro[last_state][curr_state]*e_pro[curr_state][obs[i]], last_state) \n",
    "                       for last_state in states))\n",
    "            curr_pro[curr_state] = max_pro\n",
    "            path[curr_state].append(last_sta)\n",
    "\n",
    "    # backtrack to find max path\n",
    "    max_pro = -1\n",
    "    max_path = None\n",
    "    for s in states:\n",
    "        path[s].append(s)\n",
    "        if curr_pro[s] > max_pro:\n",
    "            max_path = path[s]\n",
    "            max_pro = curr_pro[s]\n",
    "    return (max_path, max_pro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(['B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B'], 4.829216754076891e-27)\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Coding and non Coding DNA Regions\n",
    "- H state --> high GC content --> coding DNA region\n",
    "- L state --> low GC content --> non coding DNA region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = ('H', 'L')\n",
    " \n",
    "observations = ('A', 'C', 'G', 'T')\n",
    " \n",
    "start_probability = {'H': 0.5, 'L': 0.5}\n",
    " \n",
    "transition_probability = {\n",
    "   'H' : {'H': 0.5, 'L': 0.5},\n",
    "   'L' :   {'L': 0.6, 'H': 0.4},\n",
    "   }\n",
    " \n",
    "emission_probability = {\n",
    "   'H' : {'A': 0.2, 'C': 0.3, 'G': 0.3, 'T': 0.2},\n",
    "   'L'   : {'A': 0.3, 'C': 0.2, 'G': 0.2, 'T': 0.3}\n",
    "   }\n",
    "\n",
    "\n",
    "def Viterbi(obs, states, s_pro, t_pro, e_pro):\n",
    "    path = { s:[] for s in states} # init path: path[s] represents the path ends with s\n",
    "    curr_pro = {}\n",
    "    for s in states:\n",
    "        curr_pro[s] = s_pro[s]*e_pro[s][obs[0]]\n",
    "    for i in range(1, len(obs)):\n",
    "        last_pro = curr_pro\n",
    "        curr_pro = {}\n",
    "        for curr_state in states:\n",
    "            max_pro, last_sta = max(((last_pro[last_state]*t_pro[last_state][curr_state]*e_pro[curr_state][obs[i]], last_state) \n",
    "                       for last_state in states))\n",
    "            curr_pro[curr_state] = max_pro\n",
    "            path[curr_state].append(last_sta)\n",
    "\n",
    "    # backtrack to find max path\n",
    "    max_pro = -1\n",
    "    max_path = None\n",
    "    for s in states:\n",
    "        path[s].append(s)\n",
    "        if curr_pro[s] > max_pro:\n",
    "            max_path = path[s]\n",
    "            max_pro = curr_pro[s]\n",
    "    return (max_path, max_pro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(['H', 'H', 'H', 'L', 'L', 'L', 'L', 'L', 'L'], 4.251527999999999e-08)\n"
     ]
    }
   ],
   "source": [
    "obs = [char for char in \"GGCACTGAA\"]\n",
    "print(Viterbi(obs, states, start_probability, transition_probability, emission_probability))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
