{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "def loadFasta(filename):\n",
    "    \"\"\" Parses a classically formatted and possibly \n",
    "        compressed FASTA file into two lists. One of \n",
    "        headers and a second list of sequences.\n",
    "        The ith index of each list correspond.\"\"\"\n",
    "    if (filename.endswith(\".gz\")):\n",
    "        fp = gzip.open(filename, 'r')\n",
    "    else:\n",
    "        fp = open(filename, 'r')\n",
    "    # split at headers\n",
    "    data = fp.read().split('>')\n",
    "    fp.close()\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)     \n",
    "    headers = []\n",
    "    sequences = []\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        headers.append(lines.pop(0))\n",
    "        # add an extra \"+\" to make string \"1-referenced\"\n",
    "        sequences.append('+' + ''.join(lines))\n",
    "    return (headers, sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns a dictionary with the counts of each unique kmer present in a given sequence\n",
    "# there will be N-k+1 total kmers, where N = len(seq)-1 (-1 to account for the \"+\")\n",
    "# there are always 4**k total possible kmers\n",
    "\n",
    "def kmerCounts(seq, k):\n",
    "    kmerDict = {}\n",
    "    for i in range(1,len(seq)-k+1):\n",
    "        kmer = seq[i:i+k]\n",
    "        kmerDict[kmer] = kmerDict.get(kmer,0) + 1\n",
    "    return kmerDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AG': 1,\n",
       " 'GA': 1,\n",
       " 'AC': 2,\n",
       " 'CA': 2,\n",
       " 'AT': 1,\n",
       " 'TG': 1,\n",
       " 'GT': 1,\n",
       " 'TC': 1,\n",
       " 'AA': 1}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmerCounts(\"TAGACATGTCAAC\", 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmerTwice(k, string):\n",
    "    kmerDict = {}\n",
    "    for i in range(1,len(string)-k+1):\n",
    "        kmer = string[i:i+k]\n",
    "        kmerDict[kmer] = kmerDict.get(kmer,0) + 1\n",
    "    \n",
    "    hasTwice = []    \n",
    "    \n",
    "    for kmer in kmerDict:\n",
    "        if kmerDict[kmer] == 2:\n",
    "            hasTwice.append(kmer)\n",
    "            \n",
    "    return hasTwice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AC', 'CA']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmerTwice(2, \"TAGACATGTCAAC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon = {  # Maps an RNA triplet of nucelotides to a 1-letter Amino Acid Abbrevation \n",
    "    \"AAA\": 'K', \"AAG\": 'K', \"AAC\": 'N', \"AAT\": 'N',\n",
    "    \"AGA\": 'R', \"AGG\": 'R', \"AGC\": 'S', \"AGT\": 'S',\n",
    "    \"ACA\": 'T', \"ACG\": 'T', \"ACC\": 'T', \"ACT\": 'T',\n",
    "    \"ATA\": 'I', \"ATG\": 'M', \"ATC\": 'I', \"ATT\": 'I',\n",
    "    \"GAA\": 'E', \"GAG\": 'E', \"GAC\": 'D', \"GAT\": 'D',\n",
    "    \"GGA\": 'G', \"GGG\": 'G', \"GGC\": 'G', \"GGT\": 'G',\n",
    "    \"GCA\": 'A', \"GCG\": 'A', \"GCC\": 'A', \"GCT\": 'A',\n",
    "    \"GTA\": 'V', \"GTG\": 'V', \"GTC\": 'V', \"GTT\": 'V',\n",
    "    \"CAA\": 'Q', \"CAG\": 'Q', \"CAC\": 'H', \"CAT\": 'H',\n",
    "    \"CGA\": 'R', \"CGG\": 'R', \"CGC\": 'R', \"CGT\": 'R',\n",
    "    \"CCA\": 'P', \"CCG\": 'P', \"CCC\": 'P', \"CCT\": 'P',\n",
    "    \"CTA\": 'L', \"CTG\": 'L', \"CTC\": 'L', \"CTT\": 'L',\n",
    "    \"TAA\": '*', \"TAG\": '*', \"TAC\": 'Y', \"TAT\": 'Y',\n",
    "    \"TGA\": '*', \"TGG\": 'W', \"TGC\": 'C', \"TGT\": 'C',\n",
    "    \"TCA\": 'S', \"TCG\": 'S', \"TCC\": 'S', \"TCT\": 'S',\n",
    "    \"TTA\": 'L', \"TTG\": 'L', \"TTC\": 'F', \"TTT\": 'F'\n",
    "}\n",
    "\n",
    "# maps a DNA sequence to a string of 1-letter amino acid abbreviations\n",
    "\n",
    "def dna_to_protein(seq):\n",
    "    return ''.join([codon[seq[i:i+3]] for i in range(0,len(seq),3)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns the reverse complement of a given DNA sequence\n",
    "\n",
    "def revComp(dnaSeq):\n",
    "    return ''.join([{'A':'T','C':'G','G':'C','T':'A'}[base] for base in reversed(dnaSeq)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
