{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hidden Markov Models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mixed Genome Problem\n",
    "- given a set of alleles from the alphabet {A,C,G,T,H,N} (where H and N indicate probe errors) from 8 founder genomes and a 9th hybrid genome, determine the founder genome that most likely contributed to each allele in the hybrid genome\n",
    "    - i.e. determine the most likely set of mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'GenomeData.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-7d93be31e43f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"GenomeData.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m        \u001b[0;31m# break file into lines\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m     \u001b[0;31m# First line is header\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mcolumn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"owenmc\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'GenomeData.csv'"
     ]
    }
   ],
   "source": [
    "fp = open(\"GenomeData.csv\", 'r')\n",
    "data = fp.read().split('\\n')        # break file into lines\n",
    "fp.close()\n",
    "header = data.pop(0).split(',')     # First line is header\n",
    "column = header.index(\"owenmc\")\n",
    "# while (len(data[-1].strip()) < 1):  # remove extra lines\n",
    "#     data.pop()\n",
    "for i, line in enumerate(data):     # make a list from each row\n",
    "    field = line.split(',')\n",
    "    field[1] = int(field[1])        # convert position to integer\n",
    "    for base in field[2:10]:\n",
    "        if base != field[3]:\n",
    "            break\n",
    "    data[i] = field[0:10] + [base]\n",
    "    \n",
    "\n",
    "for i in range(0, 10):\n",
    "    print(\"data[%d] = %s\" % (i, data[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import exp, log10\n",
    "\n",
    "Nstates = 8\n",
    "prevpos = 1\n",
    "state = [[(float(len(data)),i) for i in range(Nstates)]]   # (log(p), PathToHere)\n",
    "for i in range(len(data)):\n",
    "    # Count expected genotypes\n",
    "    count = dict([(call, data[i][2:2+Nstates].count(call)) for call in \"ACGTHN\"])\n",
    "    # Get the target genotype at this probe\n",
    "    observed = data[i][-1]\n",
    "    # Compute emission probability, assuming 5% error rate\n",
    "    if (count[observed] == 0):\n",
    "        emission = [1.0/Nstates for j in xrange(2,2+Nstates)]   # unexpected \n",
    "    else:\n",
    "        emission = [0.99/count[data[i][j]] if data[i][j] == observed else 0.01/(Nstates - count[data[i][j]])\n",
    "                    for j in range(2,2+Nstates)]\n",
    "    # compute transition probability\n",
    "    position = data[i][1]\n",
    "    delta = position - prevpos\n",
    "    prevpos = position\n",
    "    stay = ((Nstates - 1.0)*exp(-delta/23604450.50) + 1.0)/Nstates\n",
    "    switch = (1.0 - stay)/(Nstates - 1.0)\n",
    "    # update state probailities for all paths leading to the ith state\n",
    "    path = []\n",
    "    for j in range(Nstates):\n",
    "        choices = [(log10(emission[j])+(log10(stay) if (k==j) else log10(switch))+state[-1][k][0],k)\n",
    "                   for k in range(Nstates)]\n",
    "        path.append(max(choices))   # choices is a list of tuples of (score[i], from_whence_I_arrived[i])\n",
    "    state.append(path)\n",
    "print(\"Length of paths:\", len(state))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
